plugins/out_s3/s3.c

```diff
@@ -1734,22 +1814,18 @@ static void s3_upload_queue(struct flb_config *config, void *out_context)
             s3_store_file_lock(upload_contents->upload_file);
             ctx->upload_queue_success = FLB_FALSE;
 
-            /* If retry limit was reached, discard file and remove file from queue */
-            upload_contents->retry_counter++;
-            if (upload_contents->retry_counter >= MAX_UPLOAD_ERRORS) {
-                flb_plg_warn(ctx->ins, "Chunk file failed to send %d times, will not "
-                             "retry", upload_contents->retry_counter);
-                s3_store_file_inactive(ctx, upload_contents->upload_file);
-                multipart_upload_destroy(upload_contents->m_upload_file);
+            /* If retry limit was reached, remove file from queue */
+            if (upload_contents->upload_file->failures > ctx->ins->retry_limit) {
                 remove_from_queue(upload_contents);
                 continue;
             }
```

Comment:
I really don't understand how this is managing to delete the file. From the looks of it, the code is now just removing the file from the upload queue?

Reply:
The upload_data function will handle removing chunk and dealing with upload, the S3 upload queue is a wrapper on top of that, so it just needs to remove it from its tracking queue